RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

```diff
@@ -0,0 +1,323 @@
+
+As we can see from the above diagram, we’ll separate the generation of a quantized model in PyTorch and the actual runnable quantized model on a specific backend. PyTorch will produce a reference quantized model which contains reference patterns that can be fused into quantized functions, this model will act as a unified representation of a quantized model and we do not give guarantees on either numerics or performance.
+
+Accuracy of a model on a specific backend can be emulated using reference ops as long as the numerics of the backend are well approximated by reference patterns (i.e by a sequence of dequant-fp32-quant ops). If this is not the case, then reference patterns need to be lowered to numerically accurate emulation functions for purposes of emulating accuracy.
+
+To get a model runnable on a specific backends, we will need to have an extra lowering step that transforms this reference model to a backend specific model (a model that only runs on that backend, for example: fbgemm/qnnpack). We may also transform the reference model to a backend with fake ops that simulates the numerics of the ops that run on backends.
+Backend Configurations
+A backend is a hardware or kernel library (NNPI, FBGEMM, QNNPACK etc.), each hardware/kernel library has a set of settings that can differ from the default setting, We can define them by following:
+1. **Quantization Scheme** (symmetric vs asymmetric, per-channel vs per-tensor)
+2. **Data Type** (float32, float16, int8, int8, bfloat16, etc)
+3. **Quantized (and Fused) Operators and Mapping** The quantized operators supported by the backend. For example: quantized conv2d, quantized linear etc.
+Some quantized operators may have different numerics compared to a naive (dequant - float_op - quant) implementation
+For weighted operators (conv and linear) we need to define a reference module and a mapping
+4. **QAT Module Mapping** For modules with weights, e.g. Conv2d and Linear, we need to swap them with qat (quantization aware training) module that adds fake quantization to the weights
+
+Note that this is general to all backends, not just custom backends. Current default backends in PyTorch (fbgemm/qnnpack) can be defined in the above terms as well.
+
+
+| | Fbgemm/qnnpack (supported by default)|
+| ----| ----|
+|Quantization Scheme | activation: per tensor, weight: per tensor or per channel |
+| Data Type | activation: quint8 (reduce_range for fbgemm), weight: qint8 |
+| Quantized (and Fused) Operators and Mapping | For example: nn.Conv2d → torch.ao.nn.quantized.reference.Conv2d |
+| QAT Module Mapping | Conv, Conv - Bn, Linear etc.|
+
+# Proposed APIs for Custom Backend Developers
+```python
+from enum import Enum
+# Demonstration purpose only, the api is subject to change
+class QuantizedOperatorType(Enum):
+    NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS = 0
+    OUTPUT_IS_SHARING_OBSERVER_WITH_INPUT = 1
+
+conv_module_config = {
```

Comment:
not sure whether it's the final api, but dataclasses everywhere to make things typed

Reply:
sure, will do that in the end, this is for demo only